provider/plug_block_devices.py

```diff
@@ -0,0 +1,285 @@
+
+Available methods:
+- hotplug_devices: Hot plug the devices which are defined by images by serial
+                   or parallel.
+- unplug_devices: Unplug the devices which are defined by images by serial or
+                  parallel.
+"""
+
+import logging
+import threading
+
+from avocado import TestError
+
+from virttest import utils_misc
+from virttest.qemu_devices import qdevices
+
+HOTPLUG, UNPLUG = ('hotplug', 'unplug')
+DELETED_EVENT = 'DEVICE_DELETED'
+
+
+class PlugManager(object):
+    def __init__(self, vm, params):
+        """
+        The plugging block devices manager.
+
+        :param vm: Vm object.
+        :param params: Dictionary with test parameters.
+        """
+        self.vm = vm
+        self.params = params
+        self.data_imgs = params.get("images").split()[1:]
+        self.hotplug_devs = []
+        self.unplug_devs = []
+        self.ostype = params['os_type']
+        self.plugged_disks = []
+        self.imgs = []
+        self.all_disks = set()
+        self.lock = threading.Lock()
+        self.qmp_outputs = {}
+        self.event_devs = []
+
+    def __getitem__(self, index):
+        """ Get the hot plugged disk index. """
+        if isinstance(index, slice):
+            return [self.plugged_disks[i] for i in range(*index.indices(len(self)))]
+        elif isinstance(index, int):
+            return self.plugged_disks[index]
+        else:
+            raise TestError('PlugManager indices must be integers.')
+
+    def __len__(self):
+        """ Get the len of the hot plugged disks. """
+        return len(self.plugged_disks)
+
+    def __iter__(self):
+        """ Iterate the the hot plugged disks. """
+        for disk in self.plugged_disks:
+            yield disk
+
+    def list_all_disks(self):
+        """ List all the disks. """
+        session = self.vm.wait_for_login(timeout=360)
+        if self.ostype == "linux":
+            self.all_disks = utils_misc.list_linux_guest_disks(session)
+        else:
+            self.all_disks = set(session.cmd('wmic diskdrive get index').split()[1:])
+        session.close()
+        return self.all_disks
+
+    def _create_block_devices(self, images):
+        """ Create the block devcies. """
+        del self.hotplug_devs[:]
+        for img in images:
+            img_params = self.params.object_params(img)
+            self.hotplug_devs.extend(
+                self.vm.devices.images_define_by_params(img, img_params, 'disk'))
+
+    def _plug_devices(self, action, devices, monitor):
+        """ Plug devices. """
+        for device in devices:
+            ret = getattr(
+                self.vm.devices, 'simple_%s' % action)(device, monitor)
+            self.qmp_outputs[device.get_qid()] = ret
+
+    def _check_qmp_output(self, action):
+        """ Check the output of qmp commands. """
+        if self.lock.acquire():
+            qmp_outputs = self.qmp_outputs.copy()
+            for dev_id in qmp_outputs:
```

Comment:
It is better not to change the dictionary at the same time looping it.

Reply:
Also, why not store hotplug, unplug qmp outputs separately in `self.qmp_outputs`, like `self.qmp_outputs[dev_id] = (htplg_output, unplg_output)`

Reply:
Thanks, have updated it.